doc/sphinxext/missing_references.py

```diff
@@ -0,0 +1,252 @@
+    other modules) or "<unknown>", inidcating that the sphinx application
+    cannot locate the original source file (usually because an extension
+    has injected text into the sphinx parsing engine).
+    """
+    (path, line) = get_source_line(node)
+
+    if path:
+
+        basepath = os.path.abspath(os.path.join(app.confdir, ".."))
+        path = os.path.relpath(path, start=basepath)
+
+        if path.startswith(os.path.pardir):
+            path = os.path.join("<external>", os.path.basename(path))
+
+    else:
+        path = "<unknown>"
+
+    if line:
+        line = str(line)
+    else:
+        line = ""
+
+    return "%s:%s" % (path, line)
+
+
+def save_missing_references_handler(app, exc):
+    """
+    At the end of the sphinx build, either save the missing references to a
+    JSON file. Also ensure that all lines of the existing JSON file are still
+    necessary.
+    """
+    if not app.config.missing_references_enabled:
+        # no-op when we are disabled.
+        return
+
+    json_path = os.path.join(app.confdir,
+                             app.config.missing_references_filename)
+
+    references_warnings = getattr(app.env, 'missing_reference_warnings', {})
+
+    # This is a dictionary of {(dtype,target): locations}
+    references_ignored = getattr(app.env,
+        'missing_references_ignored_references', {})
+    references_events = getattr(app.env, 'missing_reference_events', {})
+
+    # Warn about any reference which is no longer missing.
+    for (dtype, target), locations in references_ignored.items():
+        missing_reference_locations = references_events.get(
+                (dtype, target), [])
+
+        # For each ignored reference location, ensure a missing reference was
+        # observed. If it wasn't observed, issue a warning.
+        for ignored_refernece_location in locations:
+            if ignored_refernece_location not in missing_reference_locations:
+                msg = (f"Reference {dtype} {target} for "
+                       f"{ignored_refernece_location} can be removed"
+                       f" from {app.config.missing_references_filename}."
+                        "It is no longer a missing reference in the docs.")
+                logger.warning(msg,
+                    location=ignored_refernece_location,
+                    type='ref',
+                    subtype=dtype)
+
+    if app.config.missing_references_write_json:
+        _write_missing_references_json(references_warnings, json_path)
+
+
+def _write_missing_references_json(records, json_path):
+    """
+    Convert ignored references to a format which we can write as JSON
+
+    Convert from ``{(dtype, target): locaitons}`` to
+    ``{dtype: {target: locations}}`` since JSON can't serialize tuples.
+    """
+    transformed_records = defaultdict(dict)
+
+    for (dtype, target), paths in records.items():
+        paths = list(paths)
+        paths.sort()
+        transformed_records[dtype][target] = paths
```

Comment:
just `= sorted(paths)` and kill the two lines above

Reply:
Indeed... for some reason, I thought `sorted` would give me a wacky iterator and not serialize to JSON nicely. But today I learned!